gdbstub: Implement gdbstub within zephyr

ceolin opened the pull request:
This pr implements support for gdb remote protocol within Zephyr. This allows debug Zephyr using gdb through a serial port. It is possible to use network (TCP and UDP) transport instead of the serial, but this initial version has support to serial only.

The implementation is divided in three layers:

1 - The top layer that is responsible for the gdb remote protocol.
2 - An architecture specific layer responsible to write/read registers,
    set breakpoints, handle exceptions, ...
3 - A transport layer to be used to communicate with the host

The communication with GDB in the host is synchronous and the system
stops execution waiting for instructions and return its execution after
a "continue" or "step" command. The protocol has an exception that is
when the host sends a packet to cause an interruption, usually triggered
by a Ctrl-C. This implementation ignores this instruction though.

There is one simple hello world sample that configures qemu to have two serial ports, one used for console and other that is exposed in localhost:1234 that is used to debug.

Commit:
debug: x86: Add gdbstub for X86

It implements gdb remote protocol to talk with a host gdb during the
debug session. The implementation is divided in three layers:

1 - The top layer that is responsible for the gdb remote protocol.
2 - An architecture specific layer responsible to write/read registers,
    set breakpoints, handle exceptions, ...
3 - A transport layer to be used to communicate with the host

The communication with GDB in the host is synchronous and the systems
stops execution waiting for instructions and return its execution after
a "continue" or "step" command. The protocol has an exception that is
when the host sends a packet to cause an interruption, usually triggered
by a Ctrl-C. This implementation ignores this instruction though.

This initial work supports only X86 using uart as backend.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
samples: debug: Add a simple gdbstub sample

Sample example built with gdbstub enabled. Two serials are used, one for
normal output and another one that is used talk with gdb in the host
machine.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

andrewboie commented on arch/x86/core/ia32/gdbstub.c:
```diff
@@ -0,0 +1,165 @@
+
+	switch (vector) {
+	case 0:
+		exception = 8;
```
We have macros for all the exception IDs in arch/x86/include/kernel_arch_data.h please consider using them instead of hard-coded values here.

andrewboie replied:
also need some kind of header defining the values you are setting to. in general avoid raw numerical values as much as we can in code

andrewboie commented on arch/x86/core/ia32/gdbstub.c:
```diff
@@ -0,0 +1,165 @@
+ * Currently we just handle vectors 1 and 3 but lets keep it generic
+ * to be able to notify other exceptions in the future
+ */
+static unsigned int getException(unsigned int vector)
```
we don't use CamelCase in Zephyr's coding style

andrewboie commented on arch/x86/core/x86_mmu.c:
```diff
@@ -32,9 +32,18 @@ BUILD_ASSERT(PHYS_RAM_ADDR + PHYS_RAM_SIZE - 1ULL <=
+
 MMU_BOOT_REGION(&_image_text_start, &_image_text_size,
-		Z_X86_MMU_US);
+		MMU_TEXT_FLAGS);
```
why is text read/write? is it because we need to inject trap instructions?

ceolin replied:
yep !

andrewboie commented on include/debug/gdbstub.h:
```diff
@@ -0,0 +1,25 @@
+ * properly initialized. Be aware that different backends can have
+ * different initialization requirements.
+ */
+void z_gdb_start(void);
```
this doesn't appear to be an application-facing API.
Can we define in kernel/include/ instead?

ceolin replied:
Maybe, I was not sure where starting to debug. Initially I let it to the application start debug itself. My doubt is because different backends can have different initialization requirements. Don't know if there is a case that needs to expose it though,

ceolin replied:
Looking it again, other debug functionalities live out of the kernel, like coverage and tracing. IMHO is more consistent keep this way. Will remove the `z_` namespace though.

andrewboie commented on include/arch/gdbstub.h:
```diff
@@ -0,0 +1,60 @@
+ *
+ * This function is called by @c z_gdb_start()
+ */
+void arch_gdb_start(void);
```
these arch_ prototypes need to be moved to the arch interface headers. either kernel/include/kernel_arch_interface.h (preferred) or include/sys/arch_interface.h

ceolin replied:
not sure about it, it is not kernel API.  `sys/arch_interface.h` says:
```
Any public kernel APIs that are implemented as inline functions and need to
call architecture-specific API so will have the prototypes for the
architecture-specific APIs here. 
```

ceolin replied:
if I move these things to `kernel_arch_interface.h` I will need either move gdbstub code to kernel or include kernel internal headers to that subsystem.

andrewboie replied:
you have to put these in the arch interface headers.

Either in

`kernel/include/kernel_arch_interface.h` (preferred)
or
`include/sys/arch_interface.h`

You probably want the latter.

ceolin replied:
can you be a little bit more specific ? are you proposing just move the API declaration to `include/sys/arch_interface.h` and keep the header `include/arch/gdbstub.h` that will be responsible to include architecture specific header `include/arch/x86/ia32/gdbtub.h` ? or get rid of `include/sys/arch_interface.h` and include `include/arch/x86/ia32/gdbstub.h` directly in `include/sys/arch_interface.h` ? I'm ok with the move, it is just looking a little bit an alien there though.

andrewboie commented on include/arch/gdbstub.h:
```diff
@@ -0,0 +1,60 @@
+ * ARCH_GDB_NUM_REGISTERS is architecure specific and
+ * this symbol must be defined in architecure specific header
+ */
+#define GDB_NUM_REGISTERS ARCH_GDB_NUM_REGISTERS
```
please see how other ARCH_ macros are documented in include/sys/arch_interface.h

andrewboie commented on arch/x86/core/ia32/excstub.S:
```diff
@@ -121,6 +121,13 @@ SECTION_FUNC(TEXT, _exception_enter)
 #ifdef CONFIG_USERSPACE
 2:
+#endif
+#ifdef CONFIG_GDB_SUPPORT
```
can we move this earlier, perhaps right next to where the co-op registers are pushed around like 104? they are only needed for debug tools as well.

ceolin replied:
I changed, but the reason I put there was because there is some arithmetic based on $esp and as these register are conditionally pushed to stack it would require some extra `ifdef`. Solved the problem using a temporary variable

andrewboie commented on arch/x86/core/ia32/gdbstub.c:
```diff
@@ -0,0 +1,165 @@
+#include <ia32/exception.h>
+#include <inttypes.h>
+#include <debug/gdbstub.h>
+
```
a lot of this code will also work on x86_64, consider moving this file up to arch/x86/core although I suppose we can do that in a later PR once x86_64 is supported

andrewboie commented on arch/x86/core/ia32/gdbstub.c:
```diff
@@ -0,0 +1,165 @@
+#include <debug/gdbstub.h>
+
+
+extern int z_gdb_main_loop(struct gdb_ctx *ctx, bool start);
```
this must be in a header somewhere. probably some private header space for the kernel part of the GDB stub.

ceolin replied:
will make it private

ceolin replied:
I see more this pattern than having private headers. At least in this region. Where exactly you propose put this private header ? Together with the architecture specific code ? I'm sending a new version without change it because it is not clear to me that is the best thing and I want to address the other items + some extra things.

dcpleung replied:
Maybe add this into `include/debug/gdbstub.h`? Other architectures (hopefully) will have GDB stub in the future so it might be a good idea to define some common function prototypes as a starting point.

andrewboie replied:
I suggest `kernel/include/kernel_internal.h` since this API only gets invoked from the `arch/` code and not from the application. I don't care too much where it goes, but we're not going to have random extern declarations in C code

andrewboie commented on include/arch/gdbstub.h:
```diff
@@ -0,0 +1,60 @@
+ * @param addr Memory address to be read
+ * @param val [out] Memory contents
+ */
+void arch_gdb_mem_readb(uintptr_t addr, uint8_t *val);
```
quick question: under what circumstances do we need this abstraction? the current implementations just read/write memory which we could just do directly from kernel code

ceolin replied:
not now, but when we have virtual memory and debugging user threads we will need to translations afaik. I know that a lot of things will need to adapt to support it but at least the interfaces I think won't be necessary.

andrewboie replied:
do you have some additional detail about this?

my plans in this area were, if we look at memory in a GDB session that triggers a pagefault due to a swapped page, to allow the pagefault handler to do its thing and then retry. all of this should be transparent. maybe there's something I'm not considering? examples in other OS implementations might be helpful

ceolin replied:
that should work, will remove it from arch layer and if need we change it later.

andrewboie replied:
the current plan for virtual memory is a single view of RAM for the entire system from the kernel's perspective, and not independent address spaces, so we should be good with how this is currently.

tyrantlizard77 reviewed (changes requested):
One final note: I think this needs some tests. We have a sample but I'd like this better if we had some test apparatus that proves this works in a CI context. May require some creativity in setting up a harness where we instantiate QEMU with gdb over a UART session. I feel we need to write this -- the old x86 GDB stub implementation bit-rotted into uselessness over the years we had it in the kernel because nothing tested it in CI.

ceolin commented:
> One final note: I think this needs some tests. We have a sample but I'd like this better if we had some test apparatus that proves this works in a CI context. May require some creativity in setting up a harness where we instantiate QEMU with gdb over a UART session. I feel we need to write this -- the old x86 GDB stub implementation bit-rotted into uselessness over the years we had it in the kernel because nothing tested it in CI.

Yep, agree. Also I will document it, not only the API. Is it ok address it in another pr ? The other comments I'll go through and check/fix.

andrewboie commented:
> Yep, agree. Also I will document it, not only the API. Is it ok address it in another pr ? 

My general preference is, whenever feasible, to introduce tests and user documentation in the original PR.

ceolin commented:
> > Yep, agree. Also I will document it, not only the API. Is it ok address it in another pr ?
> 
> My general preference is, whenever feasible, to introduce tests and user documentation in the original PR.

Asked it because I am still thinking about how to test it and *may* be necessary add some machinery. It is possible to script gdb and test it, but integrate it CI is something that I still need to look. But ok.

nashif commented:
> the old x86 GDB stub implementation bit-rotted into uselessness over the years

the old stub never worked in the first place, it was just imported as is and was originally from "viper".

andrewboie commented:
> 
> 
> > the old x86 GDB stub implementation bit-rotted into uselessness over the years
> 
> the old stub never worked in the first place, it was just imported as is and was originally from "viper".

Don't think this changes my overall point that this mechanism needs CI testing.

nashif commented on samples/debug/README.rst:
```diff
@@ -0,0 +1,31 @@
+.. _debug_sample:
```
samples/debug is too broad, please put this under samples/debug/gdbstub..

ceolin replied:
ack

nashif commented on arch/Kconfig:
```diff
@@ -40,6 +40,7 @@ config X86
 	select HAS_DTS
 	select ARCH_HAS_CUSTOM_SWAP_TO_MAIN if !X86_64
 	select CPU_HAS_MMU
+	select ARCH_HAS_GDB_SUPPORT
```
GDB_SUPPORT is too broad and might be confusion, we probably should be specific here and replace it with GDB_SERVER or GDB_STUB

nashif reviewed (changes requested):
this needs documentation...

nashif commented:
> Don't think this changes my overall point that this mechanism needs CI testing.

we agree. We just do not have the infrastructure in place to do this yet. We need to think how to do this in CI and how to load gdb and interact with a running target..

Commit:
debug: gdbstub: Send an error code when accessing invalid memory

GDB on host may ask the guest to read the a pointer contents even if
it is pointing to NULL. Send an error code when this occurs.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
x86-32: Allow set DPL value for an exception

In order to be possible to debug usermode threads need to be able
issue breakpoint and debug exceptions. To do this it is necessary to
set DPL bits to, at least, the same CPL level.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
doc: debugging: Add information about gdbstub

An overview about gdbstub and available features.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
New things:

* Fix usermode debugging. Necessary change dpl on idt
* Add documentation
* Improve the sample, renaming and add a command file automating some gdb conmmands
* Fix an issue when the host tries to read a null pointer

ceolin commented:
@andrewboie would you mind take second look ? I have replied some questions/reviews and addressed others.  Also I have added other commits on top.

nashif commented on subsys/debug/Kconfig:
```diff
@@ -281,6 +284,34 @@ config OPENOCD_SUPPORT
+	help
+	  Use serial as backenf for GDB
+
+config GDBSTUB_SERIAL_BACKEND_NAME
```
shouldn't this be outside of the choice?

ceolin replied:
does it make sense ? I mean why allows set gdbstub serial backend name if serial backend is not selected ?

dcpleung replied:
Every `config` inside `choice`-`endchoice` pair becomes a select-able option, which means there will be two options to choose in menuconfig. So this one really needs to be moved outside of choice block.

nashif commented on subsys/debug/gdbstub/gdbstub_serial.c:
```diff
@@ -0,0 +1,53 @@
+	};
+
+	if (uart_dev == NULL) {
+#ifndef CONFIG_GDBSTUB_SERIAL_BACKEND_NAME
```
here, use the default from Kconfig instead of hardcoding uart0...

ceolin replied:
what you mean ? if GDBSTUB_SERIAL_BACKEND_NAME is defined it is used otherwise I just get uart0, is there a symbol to get default serial name ?

nashif replied:
make GDBSTUB_SERIAL_BACKEND_NAME point to UART0 by default and avoid the conditional here.

ceolin replied:
ow, I see your point, indeed, much better.

nashif commented on subsys/debug/gdbstub/gdbstub_serial.c:
```diff
@@ -0,0 +1,53 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
```
Maybe have backend in the file name, it is worth it looking at how tracing and other subsystems implement backends from different transports. Long term goal would be to reuse those instead of having each subsystem (tracing, loggin, debugging, ...) use there own backend implementation of the same transport...

nashif commented on kernel/init.c:
```diff
@@ -465,6 +466,12 @@ FUNC_NORETURN void z_cstart(void)
 	z_sys_init_run_level(_SYS_INIT_LEVEL_PRE_KERNEL_1);
+
+#if defined(CONFIG_GDBSTUB) && \
+	defined(CONFIG_GDBSTUB_SERIAL_BACKEND)
```
need to think how this would be done when we have multiple backends...

ceolin replied:
Yep, don't know exactly how to proceed right now, different backends will probably have different initialization requirements

nashif commented on samples/debug/gdbstub/README.rst:
```diff
@@ -0,0 +1,31 @@
+
+.. code-block:: bash
+
+    gdb zephyr.elf
```
gdb build/zephyr/zephyr.elf just to be accurate :)

ceolin replied:
yep, can change that. I haven't done because the build directory not necessarily needs to be called `build`. But I saw other pr using that, so will follow this pattern.

nashif commented on samples/debug/gdbstub/prj.conf:
```diff
@@ -0,0 +1,5 @@
+CONFIG_GDBSTUB=y
+CONFIG_GDBSTUB_SERIAL_BACKEND_NAME="UART_1"
+CONFIG_NO_OPTIMIZATIONS=y
+CONFIG_USERSPACE=y
```
why do we depend on userspace here?

ceolin replied:
I found an issue when setting break points from user thread cause by a general fault exception due CPL value greater than DPL, so I decided keeping user thread.

andrewboie replied:
this sounds like something we should just fix, did you adjust the DPL in the IDT entry to allow ring 3 to trigger these exceptions?

ceolin replied:
yep, added a new commit doing it

dcpleung commented on arch/Kconfig:
```diff
@@ -40,6 +40,7 @@ config X86
 	select HAS_DTS
 	select ARCH_HAS_CUSTOM_SWAP_TO_MAIN if !X86_64
 	select CPU_HAS_MMU
+	select ARCH_HAS_GDBSTUB
```
Needs `if !X86_64` if it is just for 32-bit.

ceolin replied:
done, thanks !

dcpleung commented on subsys/debug/Kconfig:
```diff
@@ -4,6 +4,9 @@
 # SPDX-License-Identifier: Apache-2.0
 
 
+config ARCH_HAS_GDBSTUB
```
Wonder if this should be moved to `arch/Kconfig`? Most of `ARCH_HAS_*` are there.

andrewboie replied:
agreed

ceolin replied:
yep, have done. Thanks !

dcpleung commented on subsys/debug/gdbstub_backend.h:
```diff
@@ -0,0 +1,48 @@
+ *
+ * @return A character
+ */
+char z_gdb_getc(void);
```
For consistency, maybe `z_gdb_getchar()` since there is a `z_gdb_putchar()`?

dcpleung commented on samples/debug/gdbstub/prj.conf:
```diff
@@ -0,0 +1,5 @@
+# nothing here
```
remove?

andrewboie commented on arch/x86/core/ia32/excstub.S:
```diff
@@ -106,18 +106,32 @@ SECTION_FUNC(TEXT, _exception_enter)
+	 * Save  the current esp in eax to avoid ifdef the code
+	 * in case of gdbstub be enable
+	 */
+	leal (%esp), %eax
```
can't we push the segment registers later and remove the need for this? you could move the block of code in `#ifdef CONFIG_GDBSTUB `to right after the `CONFIG_LAZY_FPU_SHARING` block on line 159

ceolin replied:
If I not mistaken initially I had put after push `esp` to avoid it, you suggested move to where co-op where pushed. If I change it I will need to re-order the struct again. I'd like to keep it this way if you don't mind.

andrewboie replied:
no you don't put this after you push esp. pushing esp places the esf as the argument to the C code that gets called.

>  If I change it I will need to re-order the struct again. I'd like to keep it this way if you don't mind.

I don't think we should intentionally introduce inefficiencies in our exception handling code so I'd rather this just got moved. This doesn't seem like a complicated change to me?

ceolin replied:
I was referring the `esp` before interruption occurred, pushed before `CONFIG_LAZY_FPU_SHARING`, it is a parameter in `nanoEsf`. I don't mind change it, that was basically my initial approach, to be precise it was before `CONFIG_LAZY_FPU_SHARING` but it does not seem to change anything. I was just getting confuse with it. I can avoid that instruction adding more `ifdef` but it would make it more complicated to read.

andrewboie commented on arch/x86/core/ia32/gdbstub.c:
```diff
@@ -0,0 +1,156 @@
+	z_gdb_interrupt(IV_DEBUG, esf);
+}
+
+void z_gdb_break_isr(z_arch_esf_t *esf)
```
can we make this static

andrewboie commented on arch/x86/core/ia32/gdbstub.c:
```diff
@@ -0,0 +1,156 @@
+	ctx.registers[GDB_EFLAGS] |= BIT(8);
+}
+
+void z_gdb_debug_isr(z_arch_esf_t *esf)
```
can we make this static

ceolin replied:
nope, compiler generate a warning saying that

andrewboie replied:
why? is this used outside this C file? is it in a header?

ceolin replied:
I think the compiler does not understand the construction. These functions are not directly called by C code. They are handlers registered with `__EXCEPTION_CONNECT` that expands to assembly code. The compiler is generating warning saying that these functions are not called

andrewboie replied:
have you tried using `__used`?

ceolin replied:
:) there is always an attribute.

It worked, thanks for let me know

andrewboie commented on include/debug/gdbstub.h:
```diff
@@ -0,0 +1,33 @@
+ * properly initialized. Be aware that different backends can have
+ * different initialization requirements.
+ */
+void gdb_start(void);
```
who is the intended user of this function? the kernel's init.c? the application?

ceolin replied:
I was not sure, right now I just remove this API and the debugging is initialized with `SYS_INIT` and no API is necessary.

andrewboie commented on kernel/init.c:
```diff
@@ -465,6 +466,12 @@ FUNC_NORETURN void z_cstart(void)
+
+#if defined(CONFIG_GDBSTUB) && \
+	defined(CONFIG_GDBSTUB_SERIAL_BACKEND)
+	gdb_start();
```
instead of hacking this into here, just define a SYS_INIT() for it.

ceolin replied:
ok, done. I've thought about it before but was not sure if that should be treated this way.

andrewboie commented on include/arch/gdbstub.h:
```diff
@@ -0,0 +1,38 @@
+ *
+ * This function is called by @c gdb_init()
+ */
+void arch_gdb_init(void);
```
why are these still here? needs to be in either include/sys/arch_interface.h or kernel/include/kernel_arch_interface.h

andrewboie replied:
the whole point of these headers is that we don't have all these arch interface definitions scattered all over the include space/, they are in one of two possible locations

ceolin replied:
sorry maybe this comment was lost, I am confuse because this:

can you be a little bit more specific ? are you proposing just move the API declaration to include/sys/arch_interface.h and keep the header include/arch/gdbstub.h that will be responsible to include architecture specific header include/arch/x86/ia32/gdbtub.h ? or get rid of include/sys/arch_interface.h and include include/arch/x86/ia32/gdbstub.h directly in include/sys/arch_interface.h ? I'm ok with the move, it is just looking a little bit an alien there though.

andrewboie replied:
> are you proposing just move the API declaration to include/sys/arch_interface.h

yes.
I do not want any arch_ API prototypes/documentation other than in the two headers they currently reside in.

Need a docstring for ARCH_GDB_NUM_REGISTERS, everything that is required to be defined or implemented must have doxygen for benefit of people porting to new arches.

> keep the header include/arch/gdbstub.h that will be responsible to include architecture specific header include/arch/x86/ia32/gdbtub.h

Can't you just pull this in from x86's arch.h like the rest of the platform definitions?

andrewboie replied:
To be crystal clear: I absolutely care about making sure that the places where arch_ APIs are defined do not proliferate beyond the two headers they are already in, the only reason why it's currently two and not one is because of inline function needs.

andrewboie replied:
other than this issue and a request to open a GH enhancement ticket for CI-based testing (which can be assigned to someone else) this is a +1

andrewboie commented:
This also needs an enhancement GH Issue filed to implement CI testing of this mechanism
